syntax = "proto2";

package DSSE.msg;

// The server API takes a Request and returns a Result
message Request {
    oneof msg {
        Setup setup = 1;
        Search search = 2;
        Add add = 3;
    }
}

message Result {
    optional string error = 1;
    oneof msg {
        SetupResult setup = 2;
        SearchResult search = 3;
        AddResult add = 4;
    }
}

message Setup {
    message TokenPair {
        // TODO: rename counter to token for consistency with the rest of the codebase
        optional bytes Counter = 2; // hashed counter
        optional bytes FileID = 3; // encrypted fileid
    }
    repeated TokenPair L = 1;
}

message SetupResult {
}

message Search {
    optional bytes K1 = 1;
    optional bytes K2 = 2;
    optional bytes K1plus = 3;
    optional bytes K2plus = 4;
    optional bytes K1minus = 5;
}

message SearchResult {
    repeated int64 fileid = 1;
}


message Add {
    message Pair {
        optional bytes Token = 2;
        optional bytes FileID = 3;
    }
    repeated Pair L = 1;
}
message AddResult {
}
